### DISCUSSION TITLE
Discussion: lightweight dynamic

### DISCUSSION THREAD

**Discussion by alrz** [dotnet/csharplang#539] [Language Ideas]
Moved from https://github.com/dotnet/roslyn/issues/16077

This can be a lightweight compile-time alternative to `System.Dynamic.DynamicObject`,
```cs
var r = obj.$Property;
var r = obj["Property"]; // indexer is prefered over TryGetProperty
var r = obj.TryGetProperty("Property");

obj.$Property = value;
obj["Property"] = value; // indexer is prefered over TrySetProperty
obj.TrySetProperty("Property", value);

var r = obj.$Method();
var r = obj.TryInvoke("Method", Array.Empty<Argument>());

var r = obj.Method$();
var r = obj.Method(Array.Empty<Argument>());
```
Open question: should we implicitly cast the result to the target type?
```cs
string r = obj.$Property;
string r = obj["Property"] as string;
```
Related: https://github.com/dotnet/roslyn/issues/3555

**Comment by fanoI** [comment]
It surely would semplify the use Dictionaries, json and so on... I'm less sure that this have sense:

```cs
var r = obj.$Method();
```
why not simply call the method directly?

**Comment by HaloFour** [comment]
Are you also suggesting a new object model for dynamic types as an alternative to `System.Dynamic`?  Otherwise I don't get the point.  Seems like in most cases it's just an alternative syntax to using an indexer which provides no real benefits to consumption.  It's been argued that the IDE could provide some degree of autocomplete, but I don't see why that wouldn't also be the case for supplying an argument to an indexer for a known type.

**Comment by alrz** [comment]
@HaloFour 

>Are you also suggesting a new object model for dynamic types as an alternative to System.Dynamic? 



`System.Dynamic.DynamicObject` depends on DLR, but this would use the statically resolved members.

@fanoI 

>why not simply call the method directly?

When you want to handle calls as "messages" and you don't have a fixed scheme this'll help. `obj.Method$();` variant allows you to call an existing method but with a dynamic argument list which would be passed in the form of `new Argument(name: .. , value: ... )`. These are possible using `dynamic` or a custom `System.Dynamic.DynamicObject` but, as I said, it depends on DLR and involves run-time cost.

**Comment by HaloFour** [comment]
@alrz 

If it's relying on either an indexer or some `Try` method then it's not statically resolved, at least not anymore than `dynamic` currently is.  It is certainly more lightweight than `DynamicObject` and `IDynamicMetaObjectProvider` but it completely duplicates that functionality.

**Comment by GeirGrusom** [comment]
Sounds a little like [D's forwarding operator](https://dlang.org/spec/operatoroverloading.html#dispatch).

**Comment by alrz** [comment]
>If it's relying on either an indexer or some Try method then it's not statically resolved

I meant the Try methods themselves. Even though Try methods in a DynamicObject are somehow known to the DLR, the compiler needs to emit dynamic call-sites for `dynamic` objects.
